src/utils/process.rs

```diff
@@ -1,32 +1,108 @@
 }
 
 pub fn git_blame_filename_extension() -> Option<String> {
+    calling_process_cmdline(blame::guess_git_blame_filename_extension)
+}
+
+mod blame {
+    use super::*;
+
+    // Skip all arguments starting with '-' from `args_it`. Also skip all arguments listed in
+    // `skip_this_plus_parameter` plus their respective next argument.
+    // Keep all arguments once a '--' is encountered.
+    // (Note that some arguments work with and without '=': '--foo' 'bar' / '--foo=bar')
+    fn skip_uninteresting_args<'a, 'b, ArgsI, SkipI>(
+        mut args_it: ArgsI,
+        skip_this_plus_parameter: SkipI,
+    ) -> Vec<&'a str>
+    where
+        ArgsI: Iterator<Item = &'a str>,
+        SkipI: Iterator<Item = &'b str>,
+    {
+        let arg_follows_space: HashSet<&'b str> = skip_this_plus_parameter.into_iter().collect();
+
+        let mut result = Vec::new();
+        loop {
+            match args_it.next() {
+                None => break result,
+                Some("--") => {
+                    result.extend(args_it);
+                    break result;
+                }
+                Some(arg) if arg_follows_space.contains(arg) => {
+                    let _skip_parameter = args_it.next();
+                }
+                Some(arg) if !arg.starts_with('-') => {
+                    result.push(arg);
+                }
+                Some(_) => { /* skip: --these -and --also=this */ }
+            }
+        }
+    }
+
+    pub fn guess_git_blame_filename_extension(args: &[String]) -> ProcessArgs<String> {
+        {
+            let mut it = args.iter();
+            match (it.next(), it.next()) {
+                // git blame or git -C/-c etc. and then (maybe) blame
+                (Some(git), Some(blame))
+                    if git.contains("git") && (blame == "blame" || blame.starts_with('-')) => {}
+                _ => return ProcessArgs::OtherProcess,
+            }
+        }
+
+        let args = args.iter().skip(2).map(|s| s.as_str());
+
+        // See git(1) and git-blame(1). Some arguments separate their parameter with space or '=', e.g.
+        // --date 2015 or --date=2015.
+        let git_blame_options_with_parameter =
+            "-C -c -L --since --ignore-rev --ignore-revs-file --contents --reverse --date";
+
+        match skip_uninteresting_args(args, git_blame_options_with_parameter.split(' '))
+            .last()
+            .and_then(|&s| s.split('.').last())
+            .map(str::to_owned)
+        {
+            Some(ext) => ProcessArgs::Args(ext),
+            None => ProcessArgs::ArgError,
+        }
+    }
+} // mod blame
+
+fn calling_process_cmdline<F, T>(extract_args: F) -> Option<T>
+where
+    F: Fn(&[String]) -> ProcessArgs<T>,
+{
     let mut info = sysinfo::System::new();
     let my_pid = std::process::id() as Pid;
 
     // 1) Try the parent process. If delta is set as the pager in git, then git is the parent process.
     let parent = parent_process(&mut info, my_pid)?;
 
```

Comment:
I'd like to be able to write tests with the ability to control what appears to be the parent command. What do you think about something like the following?

1. Factor out a function something like
```rust
#[cfg(not(test))]
fn get_parent_process_command() -> Option<Vec<String>> {
    let mut info = sysinfo::System::new();
    let my_pid = std::process::id() as Pid;
    let parent = parent_process(&mut info, my_pid)?;
    Some(parent.cmd().iter().map(|s| s.to_owned()).collect())
}

#[cfg(test)]
fn get_parent_process_command() -> Option<Vec<String>> {
    // TODO
    // If env var DELTA_TEST_FAKE_PARENT_PROCESS_COMMAND is set, then parse that.
}
```

I'm not familiar with testing/patching/DI techniques in Rust -- is there a better way to do this than to use an environment variable?

Reply:
I'm not familiar either, and I don't want to pick one now as they all seem quite intrusive and would also somehow have to  wrap the external `sysinfo` traits.

I added a `ProcessInterface` trait around `sysinfo::System` with a real and a mock implementation. Nothing is exported, so this only helps with local unit tests. The mock one also does not record how it is called, so only the final result can be checked, even if the wrong code path was used to create it.

Let me know what you think!

Reply:
This seems great for unit testing. I do have one comment: I think it doesn't provide a way to do the style of "end-to-end" tests that we have e.g. in [test_example_diffs.rs](https://github.com/dandavison/delta/blob/master/src/tests/test_example_diffs.rs). My env var hack would permit those I believe, so might be tempting to add that, or a better alternative if we have one? I'd quite like to input some example `git grep -p` and `git grep -W` output lines, make the parent process command appear to match, call `integration_test_utils::run_delta`, and check that delta has formatted the grep output accordingly (with styled context header in the former case, without in the latter).

Reply:
I added something less intrusive than process-global env vars in #783 (docs and names are pending), let me know if this can be used for the integration tests as you envision.

In hindsight, `ProcessInterface` was overkill since it is only good for testing the internal process model itself. Still, better than manually testing once and then letting it rot away.